src/pacmod2_node.cpp

```diff
@@ -0,0 +1,442 @@
+#include <vector>
+
+#include "pacmod2/pacmod2_node.hpp"
+
+namespace lc = rclcpp_lifecycle;
+using LNI = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface;
+using namespace std::chrono_literals;
+
+namespace pacmod2
+{
+
+constexpr std::chrono::milliseconds PACMod2Node::SEND_CMD_INTERVAL;
+constexpr std::chrono::milliseconds PACMod2Node::INTER_MSG_PAUSE;
+
+PACMod2Node::PACMod2Node(rclcpp::NodeOptions options)
+: lc::LifecycleNode("pacmod2_driver", options)
+{
+  frame_id_ = this->declare_parameter("frame_id", "pacmod");
+  dbc_major_version_ = this->declare_parameter("dbc_major_version", 3);
+
+  RCLCPP_INFO(this->get_logger(), "frame_id: %s", frame_id_.c_str());
+  RCLCPP_INFO(this->get_logger(), "dbc_major_version: %d", dbc_major_version_);
+
+  if (dbc_major_version_ != 3) {
+    RCLCPP_ERROR(
+      this->get_logger(),
+      "This driver currently only supports PACMod DBC version 3");
+    rclcpp::shutdown();
+  }
+}
+
+PACMod2Node::~PACMod2Node()
+{
+  if (pub_thread_ && pub_thread_->joinable()) {
+    pub_thread_->join();
+  }
+}
+
+LNI::CallbackReturn PACMod2Node::on_configure(const lc::State & state)
+{
+  (void)state;
+
+  pub_can_rx_ = this->create_publisher<can_msgs::msg::Frame>("can_rx", 100);
+
+  // Reports common to all platforms
+  can_pubs_[GlobalRptMsg::CAN_ID] =
+    this->create_publisher<pacmod2_msgs::msg::GlobalRpt>("global_rpt", 20);
+  can_pubs_[AccelRptMsg::CAN_ID] =
+    this->create_publisher<pacmod2_msgs::msg::SystemRptFloat>("accel_rpt", 20);
+  can_pubs_[BrakeRptMsg::CAN_ID] =
+    this->create_publisher<pacmod2_msgs::msg::SystemRptFloat>("brake_rpt", 20);
+  can_pubs_[ShiftRptMsg::CAN_ID] =
+    this->create_publisher<pacmod2_msgs::msg::SystemRptInt>("shift_rpt", 20);
+  can_pubs_[SteeringRptMsg::CAN_ID] =
+    this->create_publisher<pacmod2_msgs::msg::SystemRptFloat>("steering_rpt", 20);
+  can_pubs_[TurnSignalRptMsg::CAN_ID] =
+    this->create_publisher<pacmod2_msgs::msg::SystemRptInt>("turn_rpt", 20);
+  can_pubs_[VehicleSpeedRptMsg::CAN_ID] =
+    this->create_publisher<pacmod2_msgs::msg::VehicleSpeedRpt>("vehicle_speed_rpt", 20);
+  can_pubs_[WheelSpeedRptMsg::CAN_ID] =
+    this->create_publisher<pacmod2_msgs::msg::WheelSpeedRpt>("wheel_speed_rpt", 20);
+	 
+  pub_enabled_ = this->create_publisher<std_msgs::msg::Bool>(
+    "enabled", rclcpp::QoS(1).transient_local());
+  pub_all_system_statuses_ = this->create_publisher<pacmod2_msgs::msg::AllSystemStatuses>(
+    "all_system_statuses", 20);
+
+  sub_can_tx_ = this->create_subscription<can_msgs::msg::Frame>(
+    "can_tx", 100, std::bind(&PACMod2Node::callback_can_tx, this, std::placeholders::_1));
+
+  // Commands common to all platforms
+  can_subs_[AccelCmdMsg::CAN_ID] = std::make_pair(
+    this->create_subscription<pacmod2_msgs::msg::SystemCmdFloat>(
+      "accel_cmd", 20,
+      std::bind(&PACMod2Node::callback_accel_cmd, this, std::placeholders::_1)),
+    std::shared_ptr<LockedData>(new LockedData(AccelCmdMsg::DATA_LENGTH)));
```

Comment:
In c++14, we can use `std::make_shared<>` to create a shared pointer. It is a preferred way.
https://learn.microsoft.com/en-us/cpp/cpp/how-to-create-and-use-shared-ptr-instances?view=msvc-170

Reply:
don't want to make this change because of this:
https://stackoverflow.com/questions/70027525/undefined-reference-to-initialized-static-member-variable-with-make-shared

work around is `std::make_shared<LockedData>(+AccelCmdMsg::DATA_LENGTH)` but I think this is weird. I can make the change if others disagree.